grails-app/services/com/unifina/service/PermissionService.groovy

```diff
@@ -633,8 +562,107 @@ class PermissionService {
 		}
 	}
 
-	/** null is often a valid value (but not a valid user), and means "anonymous Permissions only" */
-	private static boolean isNotNullAndIdNotNull(userish) {
-		return userish != null && userish.id != null
+	Permission savePermissionAndSendShareResourceEmail(SecUser apiUser, Key apiKey, Operation op, String targetUsername, EmailMessage msg) {
+		SecUser userish = SecUser.findByUsername(targetUsername)
+		Permission permission = savePermission(msg.resource, apiUser, apiKey, userish, op)
+		sendEmailShareResource(op, targetUsername, msg)
+		return permission
+	}
+
+	Permission saveAnonymousPermission(SecUser apiUser, Key apiKey, Operation op, Resource resource) {
+		Object res = resource.load(apiUser, apiKey, true)
+		Permission permission = grantAnonymousAccess(apiUser, res, op)
+		return permission
+	}
+
+	private Permission savePermission(Resource resource, SecUser apiUser, Key apiKey, Userish targetUserish, Operation op) {
+		Object res = resource.load(apiUser, apiKey, true)
+		Permission permission = grant(apiUser, res, targetUserish, op)
+		return permission
+	}
+
+	private void sendEmailShareResource(Operation op, String username, EmailMessage msg) {
+		if (op == Operation.STREAM_GET || op == Operation.CANVAS_GET || op == Operation.DASHBOARD_GET) {
+			// Users with email/password registration get an email
+			// send only one email for each read/get permission
+			String content = groovyPageRenderer.render([
+				template: "/emails/email_share_resource",
+				model: [
+					sharer  : msg.from,
+					resource: msg.resourceType(),
+					name    : msg.resourceName(),
+					link    : msg.link(),
+				],
+			])
+			mailService.sendMail {
+				from: msg.from
+				to: msg.to
+				subject: msg.subject()
+				html: content
+			}
+		}
+	}
+
+	Permission savePermissionAndSendEmailShareResourceInvite(SecUser apiUser, String username, Operation op, EmailMessage msg) {
+		SignupInvite invite = SignupInvite.findByUsername(username)
+		if (!invite) {
+			invite = signupCodeService.create(username)
+			String content = groovyPageRenderer.render([
+				template: "/emails/email_share_resource_invite",
+				model: [
+					sharer  : msg.from,
+					resource: msg.resourceType(),
+					name    : msg.resourceName(),
+					invite  : invite,
+				],
+			])
+			mailService.sendMail {
+				from: msg.from
+				to: invite.username
+				subject: msg.subject()
+				html: content
+			}
+			invite.sent = true
+			invite.save(failOnError: true, validate: true)
+		}
+		Permission newPermission = savePermission(msg.resource, apiUser, null, invite, op)
+		return newPermission
+	}
+
+	Permission savePermissionAndCreateEthereumAccount(String username, SecUser grantor, Operation operation, Resource res) {
+		EthereumIntegrationKeyService ethereumIntegrationKeyService = Holders.getApplicationContext().getBean(EthereumIntegrationKeyService)
+		SecUser user = ethereumIntegrationKeyService.createEthereumUser(username)
```

Comment:
Repeating one earlier comment: the address might already be bound to a user, so we should `getOrCreate` the user instead of always creating it. (and change the name of this method to something like `savePermissionForEthereumAccount`.